crates/viewer/re_viewer/src/ui/recordings_panel.rs

```diff
@@ -152,90 +149,227 @@ fn recording_list_ui(
+        }
+    }
+}
+
+#[derive(Clone, Hash)]
+enum DatasetKind {
+    Remote(re_uri::Origin, String),
+    Local(ApplicationId),
 }
 
-fn app_and_its_recordings_ui(
+impl DatasetKind {
+    fn name(&self) -> &str {
+        match self {
+            Self::Remote(_, dataset) => dataset,
+            Self::Local(app_id) => app_id.as_str(),
+        }
+    }
+
+    fn select(&self, ctx: &ViewerContext<'_>) {
+        match self {
+            Self::Remote(origin, dataset) => {
+                ctx.command_sender()
+                    .send_system(SystemCommand::SelectRedapDataset {
+                        origin: origin.clone(),
+                        dataset: dataset.clone(),
+                    });
+                ctx.command_sender()
+                    .send_system(SystemCommand::ChangeDisplayMode(DisplayMode::RedapBrowser));
+            }
+            Self::Local(app) => {
+                ctx.command_sender()
+                    .send_system(re_viewer_context::SystemCommand::ActivateApp(app.clone()));
+                ctx.command_sender()
+                    .send_system(SystemCommand::SetSelection(Item::AppId(app.clone())));
+            }
+        }
+    }
+
+    fn item(&self) -> Option<Item> {
+        match self {
+            Self::Remote(_, _) => None,
+            Self::Local(app_id) => Some(Item::AppId(app_id.clone())),
+        }
+    }
+
+    fn is_active(&self, ctx: &ViewerContext<'_>) -> bool {
+        match self {
+            Self::Remote(origin, _dataset) => ctx
+                .store_context
+                .recording
+                .data_source
+                .as_ref()
+                .is_some_and(|source| match source {
+                    SmartChannelSource::RedapGrpcStream(endpoint) => {
+                        &endpoint.origin == origin // TODO(lucasmerlin): Also check for dataset
+                    }
+                    _ => false,
+                }),
+            Self::Local(app_id) => &ctx.store_context.app_id == app_id,
+        }
+    }
+
+    fn close(&self, ctx: &ViewerContext<'_>, dbs: &Vec<&EntityDb>) {
+        match self {
+            Self::Remote(..) => {
+                for db in dbs {
+                    ctx.command_sender()
+                        .send_system(SystemCommand::CloseStore(db.store_id()));
+                }
+            }
+            Self::Local(app_id) => {
+                ctx.command_sender()
+                    .send_system(SystemCommand::CloseApp(app_id.clone()));
+            }
+        }
+    }
+}
+
+fn dataset_and_its_recordings_ui(
     ctx: &ViewerContext<'_>,
     ui: &mut egui::Ui,
-    app_id: &ApplicationId,
+    kind: &DatasetKind,
     mut entity_dbs: Vec<&EntityDb>,
 ) {
     entity_dbs.sort_by_key(|entity_db| entity_db.recording_property::<Timestamp>());
 
-    let app_item = Item::AppId(app_id.clone());
-    let selected = ctx.selection().contains_item(&app_item);
+    // TODO(lucasmerlin): Can we select a dataset?
```

Comment:
Not yet. Soon, the redap browser should make the dataset selectable and show its TOC though.